Convert `doc.yml` workflow to be reusable

webknjaz opened the pull request:
This patch is meant to simplify the maintenance of multiple complex workflow definitions that are tied together into a single workflow later on.

It is separated out of #97533 for atomicity.

webknjaz committed:
🎨🧪 Convert `doc.yml` workflow to be reusable

This patch is meant to simplify the maintenance of multiple complex
workflow definitions that are tied together into a single workflow
later on.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -26,7 +26,7 @@ permissions:
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}-reusable
```
This collides with the “parent” workflow if it's exactly the same, which is why it was required to introduce some difference.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -60,6 +61,28 @@ jobs:
+            Misc/**
+            .github/workflows/reusable-docs.yml
+      - name: Check for docs changes
+        if: >-
```
If this step is skipped, the job output above will fall back to `false`.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -35,6 +35,7 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 10
     outputs:
+      run-docs: ${{ steps.docs-changes.outputs.run-docs || false }}
```
The part after `||` is only used as a fallback when the first part of the expression is empty (like when that step didn't even get executed).

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -60,6 +61,28 @@ jobs:
+        id: changed-docs-files
+        uses: Ana06/get-changed-files@v2.2.0
+        with:
+          filter: |
+            Doc/**
+            Misc/**
+            .github/workflows/reusable-docs.yml
```
This reimplements what was implemented through `paths:`.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -60,6 +61,28 @@ jobs:
+  check-docs:
+    name: 📝
+    needs: check_source
+    if: fromJSON(needs.check_source.outputs.run-docs)
```
The job above sets either `true` or `false` as strings, parsing that as JSON allows not having equality checks.

webknjaz commented on .github/workflows/reusable-docs.yml:
```diff
@@ -1,29 +1,8 @@
-name: Docs
+name: 📝
```
The included workflow may start adding nested identifiers in various places in the UI which is why I tend to use a short name. Emojis work great, a single-char name would work too but would look weird.
The actual job names inside remain fully labeled.

hugovk replied:
Let's keep words (re: https://github.com/python/cpython/pull/97533#discussion_r1178495428):
```suggestion
name: Docs
```

webknjaz commented on .github/workflows/reusable-docs.yml:
I made the workflow file name prefixed with `reusable-` to start a convention that would help distinguish standalone workflows from the “inclusion snippets”.

webknjaz commented on .github/workflows/reusable-docs.yml:
```diff
@@ -1,29 +1,8 @@
 
 on:
+  workflow_call:
   workflow_dispatch:
```
I kept this to make it possible to still trigger just this workflow separately, via a manual request.

webknjaz replied:
@hugovk this is the only bit I wasn't sure about — is it useful to keep this trigger at all. If not, deleting it could improve maintainability..

webknjaz commented on .github/workflows/reusable-docs.yml:
```diff
@@ -1,29 +1,8 @@
-    paths:
-    - 'Doc/**'
-    - 'Misc/**'
-    - '.github/workflows/doc.yml'
```
I reimplemented the filtering on the calling side. Dropping the `pull_request` trigger means that it won't run standalone on PRs, but it will run via inclusion into the main build workflow.

webknjaz commented on .github/workflows/reusable-docs.yml:
```diff
@@ -1,29 +1,8 @@
+name: 📝
 
 on:
+  workflow_call:
```
This bit is what allows to “include” this workflow into another.

webknjaz reviewed (commented):
@ambv I've added a few annotations below, hopefully they're helpful

hugovk commented on .github/workflows/build.yml:
```diff
@@ -60,6 +61,28 @@ jobs:
+          echo "run-docs=true" >> "${GITHUB_OUTPUT}"
+
+  check-docs:
+    name: 📝
```
Let's keep words (re: https://github.com/python/cpython/pull/97533#discussion_r1178495428):
```suggestion
    name: Docs
```

webknjaz committed:
Replace emojis in workflow and job names with text

This change reverts using a 📝 emoji as a UI-facing component within
GitHub Actions CI/CD and related web pages. It has been requested
in the following code review thread:
https://github.com/python/cpython/pull/97533#discussion_r1178495428

Co-Authored-By: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>
